fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,567 @@
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    if block_epoch == current_epoch and block_justified_checkpoint_epoch + 1 != current_epoch:
+        return False
+
+    if block_epoch != current_epoch and block_justified_checkpoint_epoch + 2 < current_epoch:
+        return False
+
+    return (
+        is_lmd_confirmed(store, confirmation_byzantine_threshold, block_root)
+        and is_ffg_confirmed(
+            store, 
+            confirmation_byzantine_threshold, 
+            confirmation_slashing_threshold, 
+            block_root
+        )
+    )
+```
+
+## Safe Block Hash
+
+This function is used to compute the value of the `safeBlockHash` field which is passed from CL to EL in the `forkchoiceUpdated` Engine API call.
+
+### Helper Functions
+
+#### `find_confirmed_block`
+
+```python
+def find_confirmed_block(store: Store,
+                         confirmation_byzantine_threshold: int,
+                         confirmation_slashing_threshold: int,
+                         block_root: Root) -> Root:
+
+    block = store.blocks[block_root]
+    current_epoch = get_current_store_epoch(store)
+
+    if compute_epoch_at_slot(block.slot) != current_epoch:
+        return store.finalized_checkpoint.root
+
+    if is_confirmed(store, confirmation_byzantine_threshold, confirmation_slashing_threshold, block_root):
+        return block_root
+    else:
+        return find_confirmed_block(store, confirmation_byzantine_threshold,
+                                    confirmation_slashing_threshold, block.parent_root)
+
+```
+
+### `get_safe_execution_payload_hash`
+
+```python
+def get_safe_execution_payload_hash(store: Store,
+                                    confirmation_byzantine_threshold: int,
+                                    confirmation_slashing_threshold: int) -> Hash32:
+    head_root = get_head(store)
+
+    confirmed_block_root = find_confirmed_block(store, confirmation_byzantine_threshold,
+                                                confirmation_slashing_threshold, head_root)
+    confirmed_block = store.blocks[confirmed_block_root]
+
+    if compute_epoch_at_slot(confirmed_block.slot) >= BELLATRIX_FORK_EPOCH:
+        return confirmed_block.body.execution_payload.block_hash
+    else:
+        return Hash32()
+```
+
+*Note*: This helper uses beacon block container extended in [Bellatrix](../specs/bellatrix/beacon-chain.md).
+
+## Confirmation Score
+
+The confirmation score for a block is the maximum adversarial percentage weight that a confirmed block can tolerate.
+This section specifies the algorithm to calculate the confirmation score of a given block.
+This under the assumption that the adversary is willing to get as much stake as possible slashed to
+prevent a block from being confirmed.
+
+### Helper Functions
+
+#### `get_one_confirmation_score`
+
+```python
+def get_one_confirmation_score(store: Store, block_root: Root) -> int:
```

Comment:
Consider prefix all low level functions where actual computations happen with `compute_`. IMO, this is more precise semantically wise
```suggestion
def compute_one_confirmation_score(store: Store, block_root: Root) -> int:
```

Reply:
Done